tmol/score/atom_type_dependent_term.py

```diff
@@ -29,13 +33,54 @@ def __init__(self, param_db: ParameterDatabase, device: torch.device):
         self.atom_type_index = atom_type_resolver.index
         self.np_is_hydrogen = atom_type_resolver.params.is_hydrogen.cpu().numpy()
         self.np_is_heavyatom = numpy.logical_not(self.np_is_hydrogen)
+
+        self.atom_name_index = OrderedDict()
+        self.atom_unique_id_index = OrderedDict()
         self.device = device
 
+    def get_atom_unique_id_name(self, block_name, atom_name):
+        return "UNIQUE_ID:" + block_name + ":" + atom_name
+
+    def get_atom_wildcard_id_name(self, block_name, atom_name):
+        return "WILDCARD_ID:" + atom_name
+
+    def add_atom_index(self, block_type):
+        atom_names = [x.name for x in block_type.atoms]
+
+        unique_id_indices = []
+        name_indices = []
+        for atom_name in atom_names:
+            unique_id = self.get_atom_unique_id_name(block_type.name3, atom_name)
+            self.atom_unique_id_index[unique_id] = len(self.atom_unique_id_index)
+            unique_id_indices.append(self.atom_unique_id_index[unique_id])
+
+            wildcard_id = self.get_atom_wildcard_id_name(block_type.name3, atom_name)
+            if wildcard_id not in self.atom_unique_id_index:
+                self.atom_unique_id_index[wildcard_id] = len(self.atom_unique_id_index)
+            name_indices.append(self.atom_unique_id_index[wildcard_id])
+
+        bt_atom_name_index = numpy.full((len(atom_names)), -1, dtype=numpy.int32)
+        bt_atom_unique_id_index = numpy.full((len(atom_names)), -1, dtype=numpy.int32)
+
+        for index, unique_id_index in enumerate(unique_id_indices):
+            bt_atom_unique_id_index[index] = unique_id_index
+
+        for index, name_index in enumerate(name_indices):
+            bt_atom_name_index[index] = name_index
+
+        setattr(block_type, "atom_unique_ids", bt_atom_unique_id_index)
+        setattr(block_type, "atom_wildcard_ids", bt_atom_name_index)
```

Comment:
How will this react to MET and MET:nterm being in the RTS together?

Reply:
Seems like possibly it is sufficient to simply check if the atom_unique_id is in the index yet and don't add it if it is. This gets the scores very close to their previous values (though still failing).

Reply:
The above actually doesn't seem sufficient because the atoms in one variant may not match the ones in another. How do we deal with this? We can't tag them with the full name because then they would have separate IDs and the params wouldn't match (unless you added entries for the nterm variants in the cartbonded DB, which seems very silly). 

Seems like maybe you want to annotate the unpatched variant first, and then for any atoms in other variants that don't exist in the original, tag them as new atoms? I'm not really sure here.